### DISCUSSION TITLE
REST API Support

### DISCUSSION THREAD

**Discussion by Rob-Negrete** [poolborges/unicenta-pos#91] [Ideas]
Is there a REST API currently? If yes, where can I find it?

If not, are you planning to develop one?  Perhaps, I can contribute a bit here.

**Comment by Rob-Negrete** [comment]
Also, I tried to run it with `mvn compile exec:java -Dexec.mainClass="com.openbravo.pos.forms.StartPOS"` from `unicenta-opos` folder but it couldn't get connection to the mysql dockerized database. 
I guess I'm missing some step. Could you get me to the right instructions in order to configure it and running locally (from source code).

Thank you!

**Comment by poolborges** [comment]
Currently nothing related with REST API

- Question 1) But what is the use case ??

PS: Please :-) Create a second issue for the second question, they are unrelated, help me to help you

**Comment by gablex** [comment]
Rest API would be useful to create a mobile app for the unicentaopos . In some organization, they use handheld android POS and creating an app would be very useful to complement the already existing desktop app. On the same line, it would be also useful for creating a webapplication.

**Comment by HarisHashim** [comment]
@gablex @Rob-Negrete  just chiming my thought. Last time I use OpenBravo POS and Unicenta OPOS. The application is desktop base and not a server.

Probably the question is whether this desktop app has a server. That used to be OpenBravo ERP stuff. Basically we can sync data to OpenBravo ERP. Not sure about what happen when the application become Unicenta OPOS.

At this point of time, since I am going to use this again and looking at the code again. My tought is more on about sync some of the data to Firebase as some kind of serverless sync. Maybe this can be our use case?

**Comment by gablex** [comment]
Great point there "server"... Here is a detailed use case. I have customers who have a huge distributed office. They have different points where customers can make orders. Currently, I am running the database from a central server and then at each checkout point I have a desktop with the unicentapos installed connecting to the central server database via LAN. This is however too rigid for them and they had requested if they can use handheld android tablet to access the POS. I had checked online and found webswing, started open source but now closed source. That worked to some extent but unfortunately not friendly nor native. User have to open the browser and follow some unnecessary steps to access it. In my mind, I thought that having a rest API run on the server could allow me develop a mobile app to serve the POS natively on the devices and possibly create a webapp rather than just the existing desktop app .

**Comment by HarisHashim** [comment]
@gablex I am not related to this repo. Just stumbled upon this version of the code in Github. Currently not using this code but move on to an advance version 5.3 in other repo.

However let me give you my analysis,

1. Unicenta OPOS and OpenBravo POS before it is a java desktop application. Initially it use a database call Derby that is not centralized. This database live in the same machine or desktop that run the POS application. 
 
2. With OpenBravo POS, they have OpenBravo ERP. So data is sent to the ERP and use case similar to your requirement is done by the ERP. Essentially OpenBravo ERP is the server.

3. Unicenta OPOS make one advancement to allow using MySQL as the database backend. MySQL can be hosted and running on the desktop like Derby. More importantly it can also  be hosted in a server somewhere on the cloud. Since it is on the cloud, people begin to point many of the desktop POS app to one MySQL server. For instance to have many branches using one central database. Hence this is a work around of centralizing many desktop POS terminal without the need of REST API.

For your use case, maybe you can have a look at OpenBravo website. However probably they have upgrade their POS solution to something else. I am not sure since I have not play around with OpenBravo POS and OpenBravo ERP for more than 10 years 😅.

If I were to do this requirement. I will write a server web application that get data from the centralized MySQL database that implement REST API. Tablet or phone will have mobile app client that use the REST API in tandem with the web application. 

Take note that Unicenta OPOS is a desktop application. It can never be the "web application" that I mentioned above. Hence that is why question about REST api for Unicent OPOS is a weird question.

With that said. There are many ways to implement that requirement of taking order or doing POS operation on tablet/phone.

A. For instance if your POS terminal database is not centralized on cloud. Meaning database is running in a desktop. I can just write web application and host it in that desktop.  Access that web application using tablet/phone through local network or WIFI using browser on tablet/phone. This make more sense since that solution will still run if shop is without internet.

OR

B. If internet access is guaranteed. What I will do is just sync data from MySQL database to Firebase. Firebase is a Google cloud service using what we call serverless architecture. Use that as a backend for Android app that perform your requirement.

These are all out of scope for Unicenta OPOS and just ways to use the database that Unicenta OPOS is using to store data.

**Reply by poolborges** [reply]
@HarisHashim 

> @gablex I am not related to this repo. Just stumbled upon this version of the code in Github. Currently not using this code but move on to an advance version 5.3 in other repo.
> 
> However let me give you my analysis,
> 
>     1. Unicenta OPOS and OpenBravo POS before it is a java desktop application. Initially it use a database call Derby that is not centralized. This database live in the same machine or desktop that run the POS application.
> 
>     2. With OpenBravo POS, they have OpenBravo ERP. So data is sent to the ERP and use case similar to your requirement is done by the ERP. Essentially OpenBravo ERP is the server.
> 
>     3. Unicenta OPOS make one advancement to allow using MySQL as the database backend. MySQL can be hosted and running on the desktop like Derby. More importantly it can also  be hosted in a server somewhere on the cloud. Since it is on the cloud, people begin to point many of the desktop POS app to one MySQL server. For instance to have many branches using one central database. Hence this is a work around of centralizing many desktop POS terminal without the need of REST API.
> 
> 
> For your use case, maybe you can have a look at OpenBravo website. However probably they have upgrade their POS solution to something else. I am not sure since I have not play around with OpenBravo POS and OpenBravo ERP for more than 10 years 😅.
> 
> If I were to do this requirement. I will write a server web application that get data from the centralized MySQL database that implement REST API. Tablet or phone will have mobile app client that use the REST API in tandem with the web application.
> 
> Take note that Unicenta OPOS is a desktop application. It can never be the "web application" that I mentioned above. Hence that is why question about REST api for Unicent OPOS is a weird question.
> 
> With that said. There are many ways to implement that requirement of taking order or doing POS operation on tablet/phone.
> 
> A. For instance if your POS terminal database is not centralized on cloud. Meaning database is running in a desktop. I can just write web application and host it in that desktop. Access that web application using tablet/phone through local network or WIFI using browser on tablet/phone. This make more sense since that solution will still run if shop is without internet.
> 
> OR
> 
> B. If internet access is guaranteed. What I will do is just sync data from MySQL database to Firebase. Firebase is a Google cloud service using what we call serverless architecture. Use that as a backend for Android app that perform your requirement.
> 
> These are all out of scope for Unicenta OPOS and just ways to use the database that Unicenta OPOS is using to store data.

About OpenBravo POS (Desktop Version)

* OpenBravoPOS was Fork from ... (see the readme project)
* OpenBravoPOS Deprecate and removed from sourceforge, because OpenBravo now provide a Comercial Web based POS
* Openbravo ERP 
** Is not something to be used by small Company
** CLOSE DEVELPEMENT or PARTNER
** Copy itseld the Dictionary Data Model from Adampiere (Idempiere fork of Adampiere) that copy it from Compiere...


This is a fork o "Unicenta OPOS" and this code is more Advanced that Unicenta 5.3 

* **THE UNICENTA SOURCE CODE IS NOT AVAILABLE IN PUBLIC** (YOU HAVE TO PAY)
* **THE UNICENTA DEVELOPMENT IS CLOSE** 

**Currently this code support**

* MariaDB (Server based Database)
* Sqlite (Embbeded database) - you can Easly Access this data from any programming language, or push backup to cloud
* H2
* HQSLDB

Features

* Better modularity 
** Domain (Can easly used inside Springboot to build REST API)
** Hardware (Can Easly used to create a "Remote Print Server" REST API do send print data)
** App (Swing Desktop Layer)
* Use Liquidbase for Database Migration
* You can use for Script: Apache Bean Shell, Apache Groovy

Improvment

* Less bug than previous
* Better Java code (Mostly using Generic and 2000 style of coding with Object every
* Using 
* Java 17 
* Jasper Report Improvment
* 

PS: Most of use case don't need cloud, but expose in local network

**Reply by gablex** [reply]
Thank you @poolborges  for this. It gives a great insight into what is currently on this repo. I will check and see how I can possibly try using sqlite and push data to cloud e.g what @HarisHashim  suggested, the famous firebase or even supabase which I find even more friendly to me. I hope that this work is meant to be open-source in the future. I have had a big problem with great projects that are beginning opensource and then one is forced to pay a subscription fee to get the latest code. The other thing that was lacking on the parent project (unicenta opos) was a good documentation. One has to figure out everything for themselves and became even harder when one wanted to contribute to the code base.

**Reply by Altchemist** [reply]
> @HarisHashim
> 
> > @gablex I am not related to this repo. Just stumbled upon this version of the code in Github. Currently not using this code but move on to an advance version 5.3 in other repo.
> > However let me give you my analysis,
> > ```
> > 1. Unicenta OPOS and OpenBravo POS before it is a java desktop application. Initially it use a database call Derby that is not centralized. This database live in the same machine or desktop that run the POS application.
> > 
> > 2. With OpenBravo POS, they have OpenBravo ERP. So data is sent to the ERP and use case similar to your requirement is done by the ERP. Essentially OpenBravo ERP is the server.
> > 
> > 3. Unicenta OPOS make one advancement to allow using MySQL as the database backend. MySQL can be hosted and running on the desktop like Derby. More importantly it can also  be hosted in a server somewhere on the cloud. Since it is on the cloud, people begin to point many of the desktop POS app to one MySQL server. For instance to have many branches using one central database. Hence this is a work around of centralizing many desktop POS terminal without the need of REST API.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > For your use case, maybe you can have a look at OpenBravo website. However probably they have upgrade their POS solution to something else. I am not sure since I have not play around with OpenBravo POS and OpenBravo ERP for more than 10 years 😅.
> > If I were to do this requirement. I will write a server web application that get data from the centralized MySQL database that implement REST API. Tablet or phone will have mobile app client that use the REST API in tandem with the web application.
> > Take note that Unicenta OPOS is a desktop application. It can never be the "web application" that I mentioned above. Hence that is why question about REST api for Unicent OPOS is a weird question.
> > With that said. There are many ways to implement that requirement of taking order or doing POS operation on tablet/phone.
> > A. For instance if your POS terminal database is not centralized on cloud. Meaning database is running in a desktop. I can just write web application and host it in that desktop. Access that web application using tablet/phone through local network or WIFI using browser on tablet/phone. This make more sense since that solution will still run if shop is without internet.
> > OR
> > B. If internet access is guaranteed. What I will do is just sync data from MySQL database to Firebase. Firebase is a Google cloud service using what we call serverless architecture. Use that as a backend for Android app that perform your requirement.
> > These are all out of scope for Unicenta OPOS and just ways to use the database that Unicenta OPOS is using to store data.
> 
> About OpenBravo POS (Desktop Version)
> 
> * OpenBravoPOS was Fork from ... (see the readme project)
> * OpenBravoPOS Deprecate and removed from sourceforge, because OpenBravo now provide a Comercial Web based POS
> * Openbravo ERP
>   ** Is not something to be used by small Company
>   ** CLOSE DEVELPEMENT or PARTNER
>   ** Copy itseld the Dictionary Data Model from Adampiere (Idempiere fork of Adampiere) that copy it from Compiere...
> 
> This is a fork o "Unicenta OPOS" and this code is more Advanced that Unicenta 5.3
> 
> * **THE UNICENTA SOURCE CODE IS NOT AVAILABLE IN PUBLIC** (YOU HAVE TO PAY)
> * **THE UNICENTA DEVELOPMENT IS CLOSE**
> 
> **Currently this code support**
> 
> * MariaDB (Server based Database)
> * Sqlite (Embbeded database) - you can Easly Access this data from any programming language, or push backup to cloud
> * H2
> * HQSLDB
> 
> Features
> 
> * Better modularity
>   ** Domain (Can easly used inside Springboot to build REST API)
>   ** Hardware (Can Easly used to create a "Remote Print Server" REST API do send print data)
>   ** App (Swing Desktop Layer)
> * Use Liquidbase for Database Migration
> * You can use for Script: Apache Bean Shell, Apache Groovy
> 
> Improvment
> 
> * Less bug than previous
> * Better Java code (Mostly using Generic and 2000 style of coding with Object every
> * Using
> * Java 17
> * Jasper Report Improvment
> 
> PS: Most of use case don't need cloud, but expose in local network

One interesting use case for a REST API is to integrate QR code payment. This opens the door to payment integration, such as QR code. I find myself needing one, I may very well fork a version and contribute back should you be willing.

**Reply by HarisHashim** [reply]
> Features
> 
> * Better modularity
>   ** Domain (Can easly used inside Springboot to build REST API)
>   ** Hardware (Can Easly used to create a "Remote Print Server" REST API do send print data)
>   ** App (Swing Desktop Layer)
> * Use Liquidbase for Database Migration
> * You can use for Script: Apache Bean Shell, Apache Groovy
> 
> Improvment
> 
> * Less bug than previous
> * Better Java code (Mostly using Generic and 2000 style of coding with Object every
> * Using
> * Java 17
> * Jasper Report Improvment

@poolborges  Thanks for the reply. You are right about Unicenta OPOS v5.x can be buggy. For instance I am stuck on Java 11 there. I have found some old bugs that I already fix in my repo previously as well (5 to 10 years ago).

Also they change com.openbravo name space to com.unicenta, that will mess things up a bit. I will need to do some comparison between my old codes, yours and then v5 from Unicenta to see. Will circle back to this but really not sure when will have the time.

@Altchemist payment integration like QRCode should be straightforward. I am pretty sure there are already some payment integration source code as example, At most this will be consuming REST API from your payment service provider. This will definitely be in scope of unicenta-pos. My guess will be that you need to get documentation from service provider on how to consume their REST API.

With that said, in my country (Malaysia). We just need to display an image of QR Code for customer to scan and make payment using their banking app. We do not have REST API and most probably do not have such need. So maybe you can clarify more about this? I can be misunderstanding stuff since different country do things differently.

**Reply by poolborges** [reply]
This repository alreday provide QRCode Utils classe.

My Assumption you want to generate QRCode and presente that to the customer, so can read by mobile and pay using some cloud payment provider.

Use Case 1: POS with Dual Display
- Display 1: For Seller 
- Display 2: Customer

QRCode in some Portuguese Speak country

- Cabo Verde Provider: 
    Vinti4


- Portugal Provider: 
     Mbway
     PayPal 
     MULTIBANCO: Most comercial banking applications) 

From my experience you don't need REST API (Server side) to integrate QRCode

Most "QrCode Pay" solution are URL with some Information 

**QrCode Common data**

- Base Url
- Amount
- OrderId or Any ReferenceId (Same thing you save in POS and you send to confirm at your payment "Side")
- MerchantId (The Store Identification: Payment Provider will give you this
- ..

**Example:** 

````
 // Example payment data
String paymentData = "https://yourpaymentgateway.com/pay?amount=100.00&orderId=12345&merchantId=987654321";
```` 


QrCode Using utils class - Open netbeans IDE (And execute 1 of those files)

* \kriolos-opos-hardware\src\test\java\com\openbravo\pos\util\BarAndQrCodeTest.java
* \kriolos-opos-hardware\src\test\java\com\openbravo\pos\util\QRCodeTest.java

**QrCode Utils**

* (DEPRECATED) com\openbravo\pos\util\BarcodeImage.java
* (READY) com\openbravo\pos\util\GraphicCode.java

**QrCode** - Request Remote Generated

- Is possible that some Payment Provider (method, generate data on cloud, so you have to Call the provider HTTP API to get those data or QrCode image)

**Comment by poolborges** [comment]
REST API - INITIAL COMMIT 

* https://github.com/kriolos-obiz/kriolos-opos-webapi
